### ISSUE TITLE
Implement S3 connection

### ISSUE THREAD

**Issue by bplatz** [fluree/db#450]
It is highly likely we'll use S3 at least initially for Nexus, and regardless others will want it.

Some notes:
- `commit!` has a persistence guarantee, so we'll need to block before commit! returns until confirmation of a successful write. Depending on library's async APIs, this might impact how it is implemented.
- S3 writes are likely enough to fail we'll need at least some minimal retry logic and then exception reporting. IPFS likely has similar (or higher) failure rate, so ideally the same exception capturing can be used for both (as well as file conn), but if not we should create a ticket for IPFS failure/retry as to not delay having the base S3 support.
- While writing the 2 files needed for a commit (or 3, if there is a new defaultContext), it can (and should) happen in parallel as the i/o will be the bottleneck for speed. Unfortunately for IPFS we won't know the file addresses until we write, so at least IPFS must remain serial. Also, this parallelization will require a cleanup function if some writes succeed and others fail. For now we should just keep it serial, and address a parallel option in a different ticket.
- fluree/ledger has a good clean S3 library and set of base fns that hopefully can mostly get reused.
- Not sure if this can work with CLJS, if not straight forward we should make this CLJ-only and throw an exception from CLJS if someone tries to create an S3 conn. We can add a CLJS-specific ticket and address its priority independently.
- There is increasingly a number of the same/similar functions in the different conn type namespaces. It would be great to chip away at this a little bit, but anything that takes a large effort ideally should exist as its own ticket so it can get best prioritized against the other tickets.

**Comment by cap10morgan** [comment]
Agreed on reusing the S3 code from fluree/ledger, and especially the paging support in there. That was a fun little bug that didn't show up in testing because you typically weren't working with enough objects to hit the single response limit. So starting out w/ code that's resilient to that would be 👍🏻.
